lib/rubygems/safe_marshal/reader.rb

```diff
@@ -124,67 +144,123 @@ def read_false
       end
 
       def read_user_defined
-        Elements::UserDefined.new(name: read_element, binary_string: @io.read(read_integer))
+        name = read_element
+        binary_string = @io.read(read_integer)
+        Elements::UserDefined.new(name, binary_string)
       end
 
+      EMPTY_ARRAY = Elements::Array.new([].freeze).freeze
+      private_constant :EMPTY_ARRAY
+
       def read_array
-        Elements::Array.new(elements: Array.new(read_integer) do |_i|
-                                        read_element
-                                      end)
+        length = read_integer
+        return EMPTY_ARRAY if length == 0
+        elements = Array.new(length) do
+          read_element
+        end
+        Elements::Array.new(elements)
       end
 
       def read_object_with_ivars
-        Elements::WithIvars.new(object: read_element, ivars:
-          Array.new(read_integer) do
-            [read_element, read_element]
-          end)
+        object = read_element
+        ivars = Array.new(read_integer) do
+          [read_element, read_element]
+        end
+        Elements::WithIvars.new(object, ivars)
       end
 
       def read_symbol_link
-        Elements::SymbolLink.new offset: read_integer
+        offset = read_integer
+        Elements::SymbolLink.new(offset)
       end
 
       def read_user_marshal
-        Elements::UserMarshal.new(name: read_element, data: read_element)
-      end
+        name = read_element
+        data = read_element
+        Elements::UserMarshal.new(name, data)
+      end
+
+      # profiling bundle install --full-index shows that
+      # offset 6 is by far the most common object link,
+      # so we special case it to avoid allocating a new
+      # object a third of the time.
+      # the following are all the object links that
+      # appear more than 10000 times in my profiling
```

Comment:
This is interesting. Is it possible that this is "overfitting" to the data as it exists right now and this will change, or is this something intrinsically connected?